var/spack/repos/builtin/packages/amrex/package.py

```diff
@@ -241,3 +244,55 @@ def cmake_args(self):
             args.append('-DCUDA_ARCH=' + self.get_cuda_arch_string(cuda_arch))
 
         return args
+
+    # TODO: Replace this method and its 'get' use for cmake path with
+    #   join_path(self.spec['cmake'].prefix.bin, 'cmake') once stand-alone
+    #   tests can access build dependencies through self.spec['cmake'].
+    def cmake_bin(self, set=True):
+        """(Hack) Set/get cmake dependency path."""
+        filepath = join_path(self.install_test_root, 'cmake_bin_path.txt')
+        if set:
+            with open(filepath, 'w') as out_file:
+                cmake_bin = join_path(self.spec['cmake'].prefix.bin, 'cmake')
+                out_file.write('{0}\n'.format(cmake_bin))
+        else:
+            with open(filepath, 'r') as in_file:
+                return in_file.read().strip()
+
+    @run_after('build')
+    def setup_smoke_test(self):
+        """Skip setup smoke tests for AMReX versions less than 21.12."""
+        if self.spec.satisfies('@:21.11'):
+            return
+
+        self.cache_extra_test_sources(['Tests'])
+
+        # TODO: Remove once self.spec['cmake'] is available here
+        self.cmake_bin(set=True)
+
+    def test(self):
+        """Skip smoke tests for AMReX versions less than 21.12."""
+        if self.spec.satisfies('@:21.11'):
+            print("SKIPPED: Stand-alone tests not supported for this version of AMReX.")
+            return
+
+        """Perform smoke tests on installed package."""
+        # TODO: Remove/replace once self.spec['cmake'] is available here
+        cmake_bin = self.cmake_bin(set=False)
+
+        args = []
+        args.append('-S ./cache/amrex/Tests/SpackSmokeTest')
+        args.append('-DAMReX_ROOT=' + self.prefix)
+        args.extend(self.cmake_args())
+        self.run_test(cmake_bin,
+                      args,
+                      purpose='Build with same CMake version as install')
+
+        make()
```

Comment:
This fails for me.  I'm getting the following results:

```
==> [2021-11-16-11:48:56.080151] Build with same CMake version as install
==> [2021-11-16-11:48:56.080651] '/usr/tce/bin/cmake' '-S./cache/amrex/Tests/SpackSmokeTest' '-DAMReX_ROOT=/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz' '-DUSE_XSDK_DEFAULTS=ON' '-DAMReX_SPACEDIM:STRING=3' '-DBUILD_SHARED_LIBS:BOOL=OFF' '-DAMReX_MPI:BOOL=ON' '-DAMReX_OMP:BOOL=OFF' '-DXSDK_ECISION:STRING=DOUBLE' '-DXSDK_ENABLE_Fortran:BOOL=OFF' '-DAMReX_FORTRAN_INTERFACES:BOOL=OFF' '-DAMReX_EB:BOOL=OFF' '-DAMReX_LINEAR_SOLVERS:BOOL=ON' '-DAMReX_AMRDATA:BOOL=OFF' '-DAMReX_RTICLES:BOOL=OFF' '-DAMReX_PLOTFILE_TOOLS:BOOL=OFF' '-DAMReX_HDF5:BOOL=OFF' '-DAMReX_HYPRE:BOOL=OFF' '-DAMReX_PETSC:BOOL=OFF' '-DAMReX_SUNDIALS:BOOL=OFF' '-DAMReX_PIC:BOOL=OFF'
-- The C compiler identification is GNU 8.3.1
-- The CXX compiler identification is GNU 8.3.1
-- Check for working C compiler: /usr/WS1/dahlgren/releases/spack/lib/spack/env/gcc/gcc
-- Check for working C compiler: /usr/WS1/dahlgren/releases/spack/lib/spack/env/gcc/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/WS1/dahlgren/releases/spack/lib/spack/env/gcc/g++
-- Check for working CXX compiler: /usr/WS1/dahlgren/releases/spack/lib/spack/env/gcc/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE  
-- Found MPI_C: /usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib/libmpi.so (found version "3.1") 
-- Found MPI_CXX: /usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib/libmpicxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1") found components:  C CXX 
-- Configuring done
-- Generating done
-- Build files have been written to: /g/g21/dahlgren/.spack/test/helgmp6jolslzicixh5vlhg2somxnngu/amrex-develop-imekdo4
CMake Warning:
  Manually-specified variables were not used by the project:

    AMReX_FORTRAN_INTERFACES
    AMReX_LINEAR_SOLVERS
    AMReX_PLOTFILE_TOOLS
    AMReX_SUNDIALS
    USE_XSDK_DEFAULTS
    XSDK_ENABLE_Fortran
    XSDK_PRECISION


PASSED
==> [2021-11-16-11:49:01.265623] 'make' '-j16'
Scanning dependencies of target install_test
[ 33%] Building CXX object CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/AdvancePhiAllLevels.cpp.o
[ 83%] Building CXX object CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/AmrCoreAdv.cpp.o
[ 83%] Building CXX object CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/DefineVelocity.cpp.o
[ 83%] Building CXX object CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/AdvancePhiAtLevel.cpp.o
[ 83%] Building CXX object CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/main.cpp.o
[100%] Linking CXX executable install_test
/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz/lib/libamrex.a(AMReX_ParallelDescriptor.cpp.o): In functio`amrex::ParallelDescriptor::Message::count() const':
/var/tmp/dahlgren/spack-stage/spack-stage-amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz/spack-src/Src/Base/AMReX_ParallelDescriptor.cpp:267: undefined reference to `ompi_mpi_datatype_nl'
/usr/WS1/dahlgren/releases/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-8.3.1/amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz/lib/libamrex.a(AMReX_ParallelDescriptor.cpp.o): In functio`amrex::ParallelDescriptor::EndParallel()':
/var/tmp/dahlgren/spack-stage/spack-stage-amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz/spack-src/Src/Base/AMReX_ParallelDescriptor.cpp:369: undefined reference to `ompi_mpi_comm_null'
/var/tmp/dahlgren/spack-stage/spack-stage-amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz/spack-src/Src/Base/AMReX_ParallelDescriptor.cpp:360: undefined reference to `ompi_mpi_datatype_nl'
/var/tmp/dahlgren/spack-stage/spack-stage-amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz/spack-src/Src/Base/AMReX_ParallelDescriptor.cpp:361: undefined reference to `ompi_mpi_datatype_nl'
/var/tmp/dahlgren/spack-stage/spack-stage-amrex-develop-imekdo4dol3mscifwbcdyt3frhrpmufz/spack-src/Src/Base/AMReX_ParallelDescriptor.cpp:362: undefined reference to `ompi_mpi_datatype_nl'
[...snip...]
AmrCoreAdv.cpp:(.text._ZN5amrex8FabArrayINS_9FArrayBoxEE19FillBoundary_finishIS1_Li0EEEvv[_ZN5amrex8FabArrayINS_9FArrayBoxEE19FillBoundary_finishIS1_Li0EEEvv]+0x276): undefined referencto `amrex::ParallelDescriptor::Waitall(amrex::Vector<int, std::allocator<int> >&, amrex::Vector<MPI_Status, std::allocator<MPI_Status> >&)'
AmrCoreAdv.cpp:(.text._ZN5amrex8FabArrayINS_9FArrayBoxEE19FillBoundary_finishIS1_Li0EEEvv[_ZN5amrex8FabArrayINS_9FArrayBoxEE19FillBoundary_finishIS1_Li0EEEvv]+0x43b): undefined referencto `amrex::ParallelDescriptor::Waitall(amrex::Vector<int, std::allocator<int> >&, amrex::Vector<MPI_Status, std::allocator<MPI_Status> >&)'
CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/AmrCoreAdv.cpp.o: In function `amrex::FabArray<amrex::FArrayBox>::ParallelCopy_finish()':
AmrCoreAdv.cpp:(.text._ZN5amrex8FabArrayINS_9FArrayBoxEE19ParallelCopy_finishEv[_ZN5amrex8FabArrayINS_9FArrayBoxEE19ParallelCopy_finishEv]+0x20f): undefined reference to `amrex::Paralleescriptor::Waitall(amrex::Vector<int, std::allocator<int> >&, amrex::Vector<MPI_Status, std::allocator<MPI_Status> >&)'
AmrCoreAdv.cpp:(.text._ZN5amrex8FabArrayINS_9FArrayBoxEE19ParallelCopy_finishEv[_ZN5amrex8FabArrayINS_9FArrayBoxEE19ParallelCopy_finishEv]+0x401): undefined reference to `amrex::Paralleescriptor::Waitall(amrex::Vector<int, std::allocator<int> >&, amrex::Vector<MPI_Status, std::allocator<MPI_Status> >&)'
CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/AmrCoreAdv.cpp.o: In function `amrex::FabArray<amrex::FArrayBox>::PostRcvs(std::map<int, std::vector<amrex::FarrayBase::CopyComTag, std::allocator<amrex::FabArrayBase::CopyComTag> >, std::less<int>, std::allocator<std::pair<int const, std::vector<amrex::FabArrayBase::CopyComTag, std::allocator<rex::FabArrayBase::CopyComTag> > > > > const&, char*&, amrex::Vector<char*, std::allocator<char*> >&, amrex::Vector<unsigned long, std::allocator<unsigned long> >&, amrex::Vector<int, s::allocator<int> >&, amrex::Vector<int, std::allocator<int> >&, int, int) const':
AmrCoreAdv.cpp:(.text._ZNK5amrex8FabArrayINS_9FArrayBoxEE8PostRcvsERKSt3mapIiSt6vectorINS_12FabArrayBase10CopyComTagESaIS6_EESt4lessIiESaISt4pairIKiS8_EEERPcRNS_6VectorISI_SaISI_EEERNSKmSaImEEERNSK_IiSaIiEEEST_ii[_ZNK5amrex8FabArrayINS_9FArrayBoxEE8PostRcvsERKSt3mapIiSt6vectorINS_12FabArrayBase10CopyComTagESaIS6_EESt4lessIiESaISt4pairIKiS8_EEERPcRNS_6VectorISI_SaISI_ERNSK_ImSaImEEERNSK_IiSaIiEEEST_ii]+0x42e): undefined reference to `amrex::ParallelDescriptor::Message amrex::ParallelDescriptor::Arecv<char>(char*, unsigned long, int, int, int)'
CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/AmrCoreAdv.cpp.o: In function `amrex::FabArray<amrex::FArrayBox>::PostSnds(amrex::Vector<char*, std::allocatorhar*> > const&, amrex::Vector<unsigned long, std::allocator<unsigned long> > const&, amrex::Vector<int, std::allocator<int> > const&, amrex::Vector<int, std::allocator<int> >&, int)':
AmrCoreAdv.cpp:(.text._ZN5amrex8FabArrayINS_9FArrayBoxEE8PostSndsERKNS_6VectorIPcSaIS4_EEERKNS3_ImSaImEEERKNS3_IiSaIiEEERSE_i[_ZN5amrex8FabArrayINS_9FArrayBoxEE8PostSndsERKNS_6VectorIPcIS4_EEERKNS3_ImSaImEEERKNS3_IiSaIiEEERSE_i]+0xda): undefined reference to `amrex::ParallelDescriptor::Message amrex::ParallelDescriptor::Asend<char>(char const*, unsigned long, int, intint)'
CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/AmrCoreAdv.cpp.o: In function `amrex::FabArray<amrex::FArrayBox>::FillBoundary_test()':
AmrCoreAdv.cpp:(.text._ZN5amrex8FabArrayINS_9FArrayBoxEE17FillBoundary_testEv[_ZN5amrex8FabArrayINS_9FArrayBoxEE17FillBoundary_testEv]+0x4d): undefined reference to `amrex::ParallelDescptor::Test(amrex::Vector<int, std::allocator<int> >&, int&, amrex::Vector<MPI_Status, std::allocator<MPI_Status> >&)'
CMakeFiles/install_test.dir/cache/amrex/Tests/Amr/Advection_AmrCore/Source/main.cpp.o: In function `main':
main.cpp:(.text+0x5b): undefined reference to `amrex::Initialize(int&, char**&, bool, int, std::function<void ()> const&, std::ostream&, std::ostream&, void (*)(char const*))'
collect2: error: ld returned 1 exit status
make[2]: *** [install_test] Error 1
make[1]: *** [CMakeFiles/install_test.dir/all] Error 2
make: *** [all] Error 2
```

Reply:
The issue appears to be that the test finds the wrong MPI,  `Found MPI_CXX: /usr/tce/packages/mvapich2/mvapich2-2.3-intel-19.0.4/lib/libmpicxx.so`.  It finds an Intel compiled MPI, not gcc.  How can we make sure the test and install use the same MPI?

Reply:
Is it possible that the Spack smoke tests do not have the environment of builds (`setup_build_environment`) set up @tldahlgren?

Reply:
I'll have to double check.  I *thought* I'd added both but at least the `setup_run_environment`.

Reply:
I am not sure we really need to test again that we can build in test/smoke-tests. Instead, I would say we should pick a binary, e.g., of a unit test here and check _if can execute_.

The CI coverage that we can _build dependencies_ of AMReX (which is what `install_test` mainly does) is already covered: by simply building the dependencies affected by a new AMReX package version. (Spack has build-CI for that via E4S.)

Reply:
Does look like only setting `setup_run_environments` at this point.  Thinking if `test_requires_compiler` then we should/could add `setup_build_environment`.  I'll experiment with that.

Reply:
I'm not seeing either defined in the package so doesn't help in this case when I do add `setup_run_environment`.

Reply:
This isn't the change that is needed to get the test working for me.